### DISCUSSION TITLE
Accessing login page from both KC_HOSTNAME and localhost

### DISCUSSION THREAD

**Discussion by lucacavazzana** [keycloak/keycloak#21822] [Q&A]
Hi,
I recently updated the login of our Single Page Application, moving from a custom page (interacting directly with the KC API in the backend) to Keycloak's login page (using _keycloak-js_). This required configuring `KC_HOSTNAME_URL` to make KC aware of the endpoint it is now exposed on.

The problem is that some users may need to access through port-forwarding, but it doesn't work with this new setup since KC's login page tries to communicate back through the external endpoint configured in `KC_HOSTNAME_URL`. Also, FE developers are used to test the SPA running it locally while connecting to API Gateway (Kong) of a remote dev cluster, and they get redirected too to the external endpoint during the login phase (with cookie-related errors).

Is it possible to configure Keycloak in a way that allows logging both through the external endpoint and localhost?

Currently, I'm using a containerized Keycloak 20 with the following configuration:
- `KC_HOSTNAME_URL=example.io/auth`
- `KC_HOSTNAME_STRICT_HTTPS=false`
- `KC_PROXY=edge`
- `KC_HOSTNAME_STRICT=false`

The SPA connects to the backend services (including Keycloak) through Kong, configured to forward the X-Forward headers.

Among the solutions, I also tried NOT setting `KC_HOSTNAME`, but that causes the other backend services, which leverage KC to authorize the incoming requests, to fail validating the identity issuer (KC's auto-discovered http://keycloak:8080 does not match the external hostname of the incoming request).

Thanks

**Comment by darius-m** [comment]
You may be able to achieve what you want by adding a mapping for your hostname to `127.0.0.1` in a local DNS resolver or the `hosts` file of your system. Do note that this will require manually changing the DNS server configuration every time you need to use one or the other. For testing purposes you could have these configurations on a different (virtual) system (e.g. in a VM or container).

**Comment by alinturbut** [comment]
Hi,

This can be achieved by enabling dynamic backchannel urls and using Forwarded headers when connecting through local / private networks. You can find more information about this here: https://www.keycloak.org/server/hostname
